app/file.go

```diff
@@ -421,6 +431,379 @@ func (a *App) DoUploadFile(now time.Time, rawTeamId string, rawChannelId string,
 	return info, err
 }
 
+// UploadFileTask encapsulates the file upload functionality. To upload a file
+// one should use App.NewUploadFileTask(), set optional parameters as needed,
+// then call its Do() method to perform the upload.
+type UploadFileTask struct {
+	//=============================================================
+	// Parameters initialized by NewUploadFileTask
+
+	TeamId    string
+	ChannelId string
+	UserId    string
+
+	// File name.
+	Name string
+
+	// Time stamp to use when creating the file.
+	Timestamp time.Time
+
+	// The value of the Content-Length http header, when available.
+	ContentLength int64
+
+	// The file data stream.
+	Input io.Reader
+
+	//=============================================================
+	// Optional parameters
+
+	// An optional, client-assigned Id field.
+	ClientId string
+
+	// If Raw, do not execute special processing for images, just upload
+	// the file.  Plugins are still invoked.
+	Raw bool
+
+	//=============================================================
+	// Internal state
+
+	buf          *bytes.Buffer
+	limit        int64
+	limitedInput io.Reader
+	teeInput     io.Reader
+	fileinfo     *model.FileInfo
+	maxFileSize  int64
+	driverName   string
+
+	// Cached image data that (may) get initialized in preprocessImage and
+	// is used in postprocessImage
+	decoded          image.Image
+	imageType        string
+	imageOrientation int
+
+	// Testing: overrideable dependency functions
+	pluginsEnvironment *plugin.Environment
+	writeFile          func(io.Reader, string) (int64, *model.AppError)
+	saveToDatabase     func(*model.FileInfo) store.StoreChannel
+}
+
+func (a *App) NewUploadFileTask(teamId, channelId, userId, name string,
+	timestamp time.Time, contentLength int64,
+	input io.Reader) *UploadFileTask {
+
+	t := &UploadFileTask{
+		TeamId:        filepath.Base(teamId),
```

Comment:
Confused by the use of `filepath` here -- none of these are file paths? Minimally, use `path` here, but I'm inclined to think we don't need to strip anything here, or should fix something upstream?

Reply:
Copied from https://github.com/mattermost/mattermost-server/blob/master/app/file.go#L425. I am a little worried about the size of this PR, and I am not 100% certain about the Id/path semantics. Happy to change but not sure how I'd track down the dependencies of the change.